### ISSUE TITLE
CLANG_ASSISTED_PARSING tries to parse markdown files

### ISSUE THREAD

**Issue by reddwarf69** [doxygen/doxygen#8784]
doxygen 1.9.2 on Fedora 34 (built from https://src.fedoraproject.org/rpms/doxygen/tree/5671b82de32b5e12b78b8f420196fda5a7272750) (same thing with the default doxygen 1.9.1 package)

Running `doxygen -g`, enabling CLANG_ASSISTED_PARSING, and running `doxygen Doxygen` causes doxygen to complain with
```
error: /home/reddwarf/tmp/ppp1/test/test.md:1:1: error: C++ requires a type specifier for all declarations [clang]
```
when the following file is used
```
Test {#mainpage}
====
```
[test.tar.gz](https://github.com/doxygen/doxygen/files/7186566/test.tar.gz)

**Comment by albert-github** [comment]
Same accounts also for Fortran and Python file (I didn't try other types).

**Comment by Bagira80** [comment]
Similar, Doxygen tries to parse `*.d` files (probably expecting them to be source for the [D language](https://dlang.org/)) and `*.m` and `*.mm` files (probably expecting them to be Objective-C source files). However, at least in my (C++) codebase they have a different meaning.

Is there maybe some way to restrict the language or file-extensions that may be parsed by (Clang assisted) Doxygen?

**Comment by albert-github** [comment]
@Bagira80
- Please also specify the **full** doxygen version used (`doxygen -v`).
- Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

- did you have a look at the doxygen setting `EXTENSION_MAPPING` (https://www.doxygen.nl/manual/config.html#cfg_extension_mapping)

**Comment by Bagira80** [comment]
Answering my own question: I must restrict the list of file-extension in the `FILE_PATTERNS` configuration value to not include these unwanted file-extensions.

However, I still would prefer if Doxygen automatically generates the full list but lets me specify some exceptions that I do not want to parse...

**Comment by Bagira80** [comment]
> * Please also specify the full doxygen version used (doxygen -v).

I now tried with Doxygen built with the latest commit from `master` (432814d75249f89b7e324f3e5653ceb5292e7139), still linking against a static libclang 13.0.0.

> * Can you please attach a, small, self contained example (source+configuration file in a tar or zip) that allows us to reproduce the problem? Please don't add external links as they might not be persistent.

You can take the example from @reddwarf69 from above and just add some more text files with extensions `.m`, `.mm` and `.d`.

That is what I did. I just duplicated the `text.md` file with these other extensions: [test.tar.gz](https://github.com/doxygen/doxygen/files/7933948/test.tar.gz)

Doxygen succeeds but prints the following errors:
```
Parsing files
Preprocessing /home/user/test/test.m...
error: /home/user/test/test.m:1:1: warning: type specifier missing, defaults to 'int' [-Wimplicit-int] [clang]
error: /home/user/test/test.m:1:5: error: expected ';' after top level declarator [clang]
Parsing file /home/user/test/test.m...
Preprocessing /home/user/test/test.mm...
error: /home/user/test/test.mm:1:1: error: C++ requires a type specifier for all declarations [clang]
error: /home/user/test/test.mm:1:7: error: expected expression [clang]
error: /home/user/test/test.mm:1:17: error: expected ';' after top level declarator [clang]
Parsing file /home/user/test/test.mm...
Reading /home/user/test/test.d...
error: /home/user/test/test.d:1:1: error: C++ requires a type specifier for all declarations [clang]
error: /home/user/test/test.d:1:7: error: expected expression [clang]
error: /home/user/test/test.d:1:17: error: expected ';' after top level declarator [clang]
Parsing file /home/user/test/test.d...
Reading /home/user/test/test.md...
error: /home/user/test/test.md:1:1: error: C++ requires a type specifier for all declarations [clang]
error: /home/user/test/test.md:1:7: error: expected expression [clang]
error: /home/user/test/test.md:1:17: error: expected ';' after top level declarator [clang]
```

If I read the documentation for [`EXTENSION_MAPPING`](https://www.doxygen.nl/manual/config.html#cfg_extension_mapping) correctly, then the clang parser should only try to parse the `text.mm` file, because it seems to be a `C/C++` extension... unless, of course, the clang parser can also parse other languages (as D or Objective-C or Markdown).  
In that case, Doxygen probably needs more fine-grained settings for telling the clang parser what sources it may parse.

**Comment by albert-github** [comment]
I've just pushed a proposed patch, pull request #9073 to limit the usage of the Clang parsing to C / Cpp language files.

**Comment by Bagira80** [comment]
> * I don't see any evidence in your `Doxyfile` of the usage of `EXTENSION_MAPPING`. In your case there should be something like:
>   ```
>   EXTENSION_MAPPING      += m=md
>   EXTENSION_MAPPING      += mm=md
>   EXTENSION_MAPPING      += d=md
>   ```
>   though this would not solve the general problem.

That was just an example. The content could have been simple text that does not need to be parsed at all.  
But then, I assume, I should have excluded `*.m`, `*.mm` and `*.d` from the `FILE_PATTERS` configuration. Correct?

---

> I've just pushed a proposed patch, pull request #9073 to limit the usage of the Clang parsing to C / Cpp language files.

I can confirm that this fixes the errors, except for the `*.mm` files, which are correctly expected to be `C/C++` files as defined in the [Doxygen documentation](https://www.doxygen.nl/manual/starting.html#default_file_extension_mapping).

**Comment by albert-github** [comment]
@Bagira80 

The code in currently the `test.mm`  is not C / Cpp code, when it is documentation markdown in this file it should be placed in a comment block.
For the exclusion of files / directories you can also use the setting `EXCLUDE_PATTERNS`.

**Comment by albert-github** [comment]
Code has been integrated in master on GitHub (please don't close the issue as this will be done at the moment of an official release).

**Comment by doxygen** [comment]
This issue was previously marked 'fixed but not released',
which means it should be fixed in doxygen version 1.9.4.
Please verify if this is indeed the case. Reopen the
issue if you think it is not fixed and please include any additional information
that you think can be relevant (preferably in the form of a self-contained example).
